plugins/file-open-save-new/core/src/main/java/org/pentaho/di/plugins/fileopensave/providers/vfs/VFSFileProvider.java

```diff
@@ -148,81 +198,111 @@ public boolean isAvailable() {
+      if ( !hasBuckets( vfsConnectionDetails ) ) {
         return null;
       }
       @SuppressWarnings( "unchecked" )
       VFSConnectionProvider<VFSConnectionDetails> temp =
         (VFSConnectionProvider<VFSConnectionDetails>) bowl
-        .getConnectionManager().getConnectionProvider( vfsConnectionDetails.getType() );
+          .getConnectionManager().getConnectionProvider( vfsConnectionDetails.getType() );
       vfsConnectionProvider = temp;
 
       vfsRoots = vfsConnectionProvider.getLocations( vfsConnectionDetails );
     } catch ( Exception e ) {
       throw new FileException( "Error getting VFS locations. Check your credentials and for connectivity." + e.getMessage(), e );
     }
-    if ( vfsRoots.isEmpty() ) {
-      throw new FileNotFoundException( file.getPath(), file.getProvider() );
+    if ( vfsRoots == null || vfsRoots.isEmpty() ) {
+      throw new FileNotFoundException( fileDomainRoot.getPath(), fileDomainRoot.getProvider() );
     }
 
-    String scheme = vfsConnectionProvider.getProtocol( vfsConnectionDetails );
     for ( VFSRoot root : vfsRoots ) {
       VFSDirectory vfsDirectory = new VFSDirectory();
       vfsDirectory.setName( root.getName() );
       vfsDirectory.setDate( root.getModifiedDate() );
       vfsDirectory.setHasChildren( true );
       vfsDirectory.setCanAddChildren( true );
-      vfsDirectory.setParent( scheme + "://" );
-      vfsDirectory.setDomain( vfsConnectionDetails.getDomain() );
-      vfsDirectory.setConnection( vfsConnectionDetails.getName() );
-      vfsDirectory.setPath( scheme + "://" + root.getName() );
+      vfsDirectory.setParent( fileDomainRoot.getPath() );
+      vfsDirectory.setPath( fileDomainRoot.getPath() + VFSFile.DELIMITER + root.getName() );
       vfsDirectory.setRoot( NAME );
       files.add( vfsDirectory );
     }
-    this.roots.put( file.getConnection(), files );
+    this.roots.put( getKey( fileDomainRoot ), files );
     return files;
   }
 
+  /**
+   * Get key for <code>file</code>
+   * @param vfsFile
+   * @return
+   */
+  protected String getKey( VFSFile vfsFile ) {
+    return vfsFile.getPath();
+  }
+
+  /**
+   * Determines if a <code>connectionDetails</code> has "buckets".
+   * @param connectionDetails
+   * @return
+   */
+  protected boolean hasBuckets( ConnectionDetails connectionDetails ) {
+    return connectionDetails instanceof VFSConnectionDetails && ( (VFSConnectionDetails) connectionDetails).hasBuckets();
+  }
+
   /**
    * @param file
    * @param filters
    * @return
    */
   @Override
   public List<VFSFile> getFiles( VFSFile file, String filters, VariableSpace space ) throws FileException {
-    if ( file.getPath().matches( DOMAIN_ROOT ) ) {
+    if (  isConnectionRoot( file ) ) {
       List<VFSFile> rootFiles = getRoot( file );
       if ( rootFiles != null ) {
         return rootFiles;
       }
     }
     FileObject fileObject;
     try {
-      fileObject = KettleVFS.getInstance( bowl )
```

Comment:
@abrahamp-hv heads up. I refactored all this logic into KettleVFSService.java in this PR.